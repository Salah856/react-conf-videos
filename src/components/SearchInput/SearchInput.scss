.root {
  padding: 1rem 2rem;
  background: white;
  font-size: 2.6rem;
  color: #404d5b;
  font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif;
  max-width: 450px;
  width: 100%;
  border: 0;
  outline: none;
  box-sizing: border-box;

  &::placeholder {
    opacity: 0.2;
  }
}
