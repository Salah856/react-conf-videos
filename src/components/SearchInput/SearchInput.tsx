/* tslint:disable:no-shadowed-variable */
import * as React from 'react'
import { compose, withHandlers, withStateHandlers } from 'recompose'

import styles from './SearchInput.scss'

type Props = {
  filterValue: string,
  placeholder?: string,
  onChange: (e: any) => void
}

type WithState = { myRef: any }

type WithStateHandlers = {
  setRef: () => WithState
  blurRef: () => WithState
}

type WithHandlers = {
  onKeyUpHandlers: (e: any) => void
}

type CombinedProps = Props & WithState & WithStateHandlers & WithHandlers

export const SearchInputInner: React.SFC<CombinedProps> = ({
  filterValue, onChange, onKeyUpHandlers, setRef, placeholder = 'Search'
}) => (
  <input
    type="text"
    ref={setRef}
    value={filterValue}
    onChange={onChange}
    className={styles.root}
    placeholder={placeholder}
    onKeyUp={onKeyUpHandlers}
  />
)

// for testing
export const blurRef = ({myRef}: WithState) => () => myRef.blur()
export const onKeyUpHandlers = ({blurRef}: WithStateHandlers) => (e: any) => {
  if (e.keyCode === 13) {
    blurRef()
  }
}

const SearchInput = compose<CombinedProps, Props>(
  withStateHandlers<WithState, WithStateHandlers, Props>(
    { myRef: null },
    {
      setRef: () => (ref) => ({ myRef: ref }),
      blurRef
    }
  ),
  withHandlers<WithState & WithStateHandlers, WithHandlers>({
    onKeyUpHandlers
  })
)(SearchInputInner)

export default SearchInput
