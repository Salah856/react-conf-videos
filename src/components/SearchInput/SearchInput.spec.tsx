import * as React from 'react'
import { shallow, mount } from 'enzyme'
import toJSON from 'enzyme-to-json'

import SearchInput, { SearchInputInner, blurRef, onKeyUpHandlers } from './SearchInput'

describe('SearchInputInner', () => {

  const shallowComp = (mockFN = jest.fn(), placeholder?: string) => {
    return shallow(
      <SearchInputInner
        filterValue=""
        onChange={jest.fn()}
        onKeyUpHandlers={mockFN}
        setRef={jest.fn()}
        myRef={null}
        blurRef={jest.fn()}
        placeholder={placeholder}
      />
    )
  }

  it('should render', () => {
    // arrange, act
    const comp = shallowComp()
    // assert
    expect(toJSON(comp)).toMatchSnapshot()
  })

  it('should call onKeyUpHandlers on onKeyUp with keyCode 13', () => {
    // arrange
    const mockFN = jest.fn()
    const comp = shallowComp(mockFN)

    // act
    comp.find('.root').simulate('keyUp', {keyCode: 13})

    // assert
    expect(mockFN).toHaveBeenCalled()
  })

  it('should set placeholder if passed', () => {
    // arrange
    const mockFN = jest.fn()
    const placeholder = 'test placeholder'

    // act
    const comp = shallowComp(mockFN, placeholder)

    // assert
    expect(toJSON(comp)).toMatchSnapshot()
  })
})

describe('SearchInput', () => {

  it('should render and contain connected props', () => {
    // arrange
    const mockFN = jest.fn()
    const comp = <SearchInput onChange={mockFN} filterValue="" />

    // act
    const wrapper = shallow(comp)

    // assert
    expect(toJSON(wrapper)).toMatchSnapshot()
    expect(wrapper.props().filterValue).toEqual('')
  })

  it('should dispatch correct actions on change handler call', () => {
    // arrange
    const mockFN = jest.fn()
    const value = 'search input'
    const comp = <SearchInput onChange={mockFN} filterValue="" />

    // act
    const wrapper = mount(comp)
    wrapper.find('input').simulate('change', {target: { value }})

    // assert
    expect(mockFN).toHaveBeenCalled()
  })
})

describe('blueRef', () => {
  it('should call the blurRef method of the element', () => {
    // arrange
    const mockBlur = jest.fn()
    const fakeElement = {myRef: {blur: mockBlur}}

    // act
    blurRef(fakeElement)()

    // assert
    expect(mockBlur).toHaveBeenCalledTimes(1)
  })
})

describe('onKeyUpHandlers', () => {
  it('should call blurRef when keyCode is 13', () => {
    // arrange
    // tslint:disable-next-line
    const blurRef = jest.fn()
    const input = {blurRef, setRef: jest.fn()}
    const mockEvent = {keyCode: 13} as React.KeyboardEvent<HTMLInputElement>

    // act
    onKeyUpHandlers(input)(mockEvent)

    // assert
    expect(blurRef).toHaveBeenCalledTimes(1)
  })

  it('should not call blurRef when keyCode is not 13', () => {
    // arrange
    // tslint:disable-next-line
    const blurRef = jest.fn()
    const input = {blurRef, setRef: jest.fn()}
    const mockEvent = {keyCode: 14} as React.KeyboardEvent<HTMLInputElement>

    // act
    onKeyUpHandlers(input)(mockEvent)

    // assert
    expect(blurRef).toHaveBeenCalledTimes(0)
  })
})
