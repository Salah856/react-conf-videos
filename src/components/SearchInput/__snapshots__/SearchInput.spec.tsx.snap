// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput should render 1`] = `
<input
  aria-label="Search videos"
  className="root"
  onChange={[MockFunction]}
  placeholder="Search"
  type="text"
  value=""
/>
`;

exports[`SearchInput should set placeholder if passed 1`] = `
<input
  aria-label="test placeholder videos"
  className="root"
  onChange={[MockFunction]}
  placeholder="test placeholder"
  type="text"
  value=""
/>
`;
